src/js/components/ServicePlanBlocks.js

```diff
@@ -0,0 +1,193 @@
+    // Stop event from propagating to window, which would dismiss the tooltip
+    // that we're trying to open.
+    event.stopPropagation();
+
+    this.handleExternalClick();
+    this.setState({selectedBlock: block});
+    this.refs[block.getID()].triggerOpen();
+  }
+
+  handleExternalClick() {
+    if (!!this.state.selectedBlock) {
+      this.refs[this.state.selectedBlock.getID()].triggerClose();
+      this.setState({selectedBlock: null});
+    }
+  }
+
+  handleBlockRestart(blockID, phaseID) {
+    ServicePlanStore.restartBlock(this.props.service.id, blockID, phaseID);
+  }
+
+  handleBlockForceComplete(blockID, phaseID) {
+    ServicePlanStore.forceCompleteBlock(this.props.service.id, blockID,
+      phaseID);
+  }
+
+  handleBlockMouseEnter(block) {
+    this.setState({hoveredBlock: block});
+  }
+
+  handleBlockMouseLeave() {
+    this.setState({hoveredBlock: null});
+  }
+
+  getUpgradeBlocks(phaseBlocks, activePhase) {
+    let blocks = phaseBlocks.getItems().map((block, blockIndex) => {
+      let {selectedBlock} = this.state;
+
+      let hasDecisionPoint = block.hasDecisionPoint();
+      let isActive = block.isInProgress();
+      let isComplete = block.isComplete();
+
+      if (isComplete && hasDecisionPoint) {
+        hasDecisionPoint = false;
+      }
+
+      let blockClassName = classNames('service-plan-modal-details-block', {
+        'has-decision-point': hasDecisionPoint,
+        'is-active': isActive,
+        'is-selected': selectedBlock && selectedBlock.getID() === block.getID(),
+        'is-complete': isComplete
+      });
+
+      let blockIcon = <IconUpgradeBlock />
+
+      if (hasDecisionPoint) {
+        blockIcon = <IconUpgradeBlockDecisionPoint />;
+      }
+
+      return (
+        <div className={blockClassName} key={blockIndex}
+          onClick={this.handleBlockClick.bind(this, block)}
+          onMouseEnter={this.handleBlockMouseEnter.bind(this, block)}
+          onMouseLeave={this.handleBlockMouseLeave}>
+          <Tooltip
+            content={this.getUpgradeBlockTooltipContent({
+              block,
+              blockIndex,
+              activePhase
+            })}
+            interactive={true}
+            ref={block.getID()}
+            stayOpen={true}
+            suppress={true}>
+            <div className="service-plan-modal-details-block-content">
+              {blockIcon}
+            </div>
+          </Tooltip>
+        </div>
+      );
+    });
+
+    return (
+      <div className="service-plan-modal-details-blocks">
+        {blocks}
+      </div>
+    );
+  }
+
+  getUpgradeBlockTooltipContent(blockDetails) {
+    let {activePhase, block, blockIndex} = blockDetails;
+    let blockOptions = [];
+    let blockLabelSeparator = ': ';
+
+    if (block.isComplete() || block.isInProgress()) {
+      blockOptions.push(
+        <a
+          className="clickable"
+          key="restart"
+          onClick={this.handleBlockRestart.bind(this, block.getID(),
```

Comment:
you got double `block.getID()`s and `activePhase.getID()`s here too if you wanted to nip that (':